File: examples/test_forward_reverse.c
#include <stdio.h>
#include <output/test_grad_forward.c>
#include <output/test_grad_reverse.c>



int main(){ 
    int size_ders = 1; 
    int hess_size_ders = 3;
    double grad_fwd[size_ders]; 
    double grad_rev[size_ders];

    double values[1] = {0.5}; 
    int num_points = 1; 
    
    compute_grad_forward(values, num_points, grad_fwd); 
    compute_grad_reverse(values, num_points, grad_rev); 
    
   FILE *fp;
   
   printf("Writing to ./output/files/compute_grad_forward.txt\n");
   fp = fopen("./output/files/compute_grad_forward.txt", "w+");
   for(int i=0;i<size_ders;i++){ 
     fprintf(fp, "%lf \n", grad_fwd[i]); 
    } 
    fclose(fp);
    
             
   printf("Writing to ./output/files/compute_grad_reverse.txt\n");             
   fp = fopen("./output/files/compute_grad_reverse.txt", "w+");
   for(int i=0;i<size_ders;i++){ 
     fprintf(fp, "%lf \n", grad_rev[i]); 
    } 
    fclose(fp);
    
}

File: examples/test_forward_reverse.py
import sys, os
import pytest
sys.path.append('../')
import acorns
c_function = "int function_test(double p, double a){ \
    int energy = p*p*a+log((p)); \
    return 0; \
}"
if not os.path.exists('output'):
    os.mkdir('output')
acorns.autodiff(c_function, 'energy', ['p'], func = 'function_test', output_filename = 'output/test_grad_forward',
       output_func = 'compute_grad_forward')
acorns.autodiff(c_function, 'energy', ['p'], func = 'function_test', reverse_diff = True, output_filename = 'output/test_grad_reverse',
       output_func = 'compute_grad_reverse')


print("\n\n%========= Run ===========%\n")

if not os.path.exists('output/obj/'):
    os.mkdir('output/obj/')
if not os.path.exists('output/files/'):
    os.mkdir('output/files/')

os.system("gcc -std=c99 -I . -O3 -ffast-math -o output/obj/test_forward_reverse_main test_forward_reverse.c ")
os.system("./output/obj/test_forward_reverse_main")